File: scala/parsers/src/main/scala/jp/co/teruuu/parser/common/PlusParser.scala
package jp.co.teruuu.parser.common

class PlusParser[T](parser: Parser[T]) extends Parser[List[T]] {

  override def parse(input: String): ParseResult[List[T]] = {
    var next = input
    var values = List.empty[T]
    var isContinue = true
    var result: ParseResult[T] = null
    while(isContinue) {
      result = parser.parse(next)
      if(result.isInstanceOf[ParseOk[_]]) {
        values = values :+ result.asInstanceOf[ParseOk[T]].value
        next = result.asInstanceOf[ParseOk[T]].next
      } else {
        isContinue = false
      }
    }
    values.length match {
      case 0 => ParseNg(result.asInstanceOf[ParseNg[_]].message, input)
      case _ => ParseOk(values, next)
    }
  }
}

File: typescript/parser/src/parser/common/plus-parser.ts
import { Parser } from './parser';
import { ParseResult, ParseSuccess, ParseFailer } from './parser-result';

export class PlusParser<T> implements Parser<T[]> {
	parser: Parser<T>;
	public constructor(parser: Parser<T>) {
		this.parser = parser;
	}

	parse(input: string): ParseResult<T[]> {
		let continueFlg = true
		let next = input
		let result = new Array<T>()
		let parseResult: ParseResult<T>
		while(continueFlg) {
			parseResult = this.parser.parse(next)
			if(parseResult instanceof ParseSuccess) {
				result.push(parseResult.value)
				next = parseResult.next
			} else if(parseResult instanceof ParseFailer) {
				continueFlg = false
			}
		}
		if(result.length == 0 && parseResult instanceof ParseFailer) {
			return new ParseFailer(parseResult.message, next)
		} else {
			return new ParseSuccess(result, next)
		}
	}
}
